pkg/metrics/bpf.go

```diff
@@ -36,55 +56,74 @@ func newbpfCollector() *bpfCollector {
 }
 
 func (s *bpfCollector) Describe(ch chan<- *prometheus.Desc) {
-	ch <- s.bpfMapsMemory
-	ch <- s.bpfProgMemory
+	prometheus.DescribeByCollect(s, ch)
 }
 
 type memoryEntry struct {
 	BytesMemlock uint64 `json:"bytes_memlock"`
 }
 
-func getMemoryUsage(typ string) (uint64, error) {
+func getBPFUsage(typ string) (bpfUsage, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 	cmd := exec.CommandContext(ctx, "bpftool", "-j", typ, "show")
 	out, err := cmd.Output()
 	if err != nil {
-		return 0, fmt.Errorf("unable to get bpftool output: %w", err)
+		return bpfUsage{}, fmt.Errorf("unable to get bpftool output: %w", err)
 	}
 
 	var memoryEntries []memoryEntry
 	err = json.Unmarshal(out, &memoryEntries)
 	if err != nil {
-		return 0, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
+		return bpfUsage{}, fmt.Errorf("unable to unmarshal bpftool output: %w", err)
 	}
 	var totalMem uint64
 	for _, entry := range memoryEntries {
 		totalMem += entry.BytesMemlock
 	}
-	return totalMem, nil
+
+	return bpfUsage{
+		count:                 uint64(len(memoryEntries)),
+		virtualMemoryMaxBytes: totalMem,
+	}, nil
 }
 
 func (s *bpfCollector) Collect(ch chan<- prometheus.Metric) {
-	mapMem, err := getMemoryUsage("map")
-	if err != nil {
-		logrus.WithError(err).Error("Error while getting BPF maps memory usage")
-	} else {
-		ch <- prometheus.MustNewConstMetric(
-			s.bpfMapsMemory,
-			prometheus.GaugeValue,
-			float64(mapMem),
-		)
-	}
+	// Avoid querying BPF multiple times concurrently
+	_, _, _ = s.sfg.Do("collect", func() (interface{}, error) {
```

Comment:
I know I suggested single flight, but I think this is an incorrect usage. You need to use the return values, otherwise only the first caller of `Collect` is going to get the values. Single flight basically acts as a concurrent memoizer.

Reply:
ohhh I see 🙈

Reply:
alright, I've made some updates which should give more sense to the implementation of the singleflight lib!